docs(functional-tests): ADR for react conversion functional tests (#16757)
mozilla/fxa: docs/adr/0043-react-functional-tests.md

- conditionally run these react tests based on a feature flag indicating the experiment is enabled for the route group

Additional strategies that have been used include:

- using conditional logic within tests to switch between react and backbone and react for a portion of the tests steps, based on feature flag configuration
- using conditional logic within tests to run through both backbone and react versions of flows.
- skip the backbone tests for full prod rollout (but there is no configuration value that can be checked)

Additional challenges:

- competing ideas have been implemented to bridge the gaps between React and Backbone in the functional tests
- the desired end state of the react functional test migration is not universally agreed upon
- Multiple react routes at different phases of rollout means that conditional logic is required to navigate between react/backbone test steps
- The tests in `react-conversion` will need to be modified again during the cleanup phase to remove all of the feature flag check and experiment parameters

### Questions

> - Are there new pattern that would simplify conversion moving forward (e.g., user-facing locators, single page POMs, route check in POMs, etc.)

Reviewer: Not sure, this might just be a consequence of how our feature flagging worked and our design decisions.

Author: That's true, the reason that some locators that were set for Backbone no longer work is because of design decisions in React. However using user-facing locators (instead of test ids or css selectors) can in many cases (but not all) be compatible for both Backbone and React. 

This is probably an overall test architecture question though.

- Do we plan to discontinue all backbone tests? If so, when should we remove vs skip?
- For tests that are not specific to the react conversion routes but that include test steps that use the converted routes, should we create new test files, or update the POM to be react/backbone agnostic?
- How should we handle test setup (account creation and signup in fixtures) that currently use the backbone POMs?

> - How could we modify our test architecture to support coverage of both backbone and react tests while a route group experiment is enabled but not fully rolled out?

Reviewer: It should be possible to force the backbone view via experiments, ie `?forceExperiment=generalizedReactApp&forceExperimentGroup=control`. If so then we might be able to just update the POM as needed.

- Would updating the locators to be backbone/react agnostic resolve the majority of failures? (currently unknown without doing the work)
- Can we modify top-level fixtures to enroll in the react experiment instead of including `forceExperiment` params in all test steps?

## Decision Drivers

1. Test Coverage

The chosen option should ideally maintain or increase coverage for React flows and avoid compromising (skipping) coverage for Backbone flows.

2. Time & Effort

The chosen option should ideally be feasable within the planned timeframe for the signin React rollout (Train 285).

3. Best Practice & Technical Debt

The chosen option should incorporate lessons learned and avoid contributing to future maintenance by:

- Using fixtures to setup and teardown stateful dependencies, such as accounts
- Using built-in user-facing locators
- Using auto-awaiting asserts
- Avoiding conditional paths
- Minimizing use of skip and fixme annotations

## Considered Options

- Option 1: Single set of POMs for React/Backbone, update existing tests
  - Option 1a: Update by Page-First Approach

>   - Option 1b: Update by Test-First Approach

Reviewer: I think this might be the easiest and least risky thing to do.

- Option 2: Status Quo - Separate POMs for React and Backbone, Migrate Tests According to [Analysis][2]

> ## Decision Outcome
> 
> Chosen option: **Option 1b, Single set of POMs for React/Backbone, update existing tests with a test-first approach**
> 
> Option 1 chosen over 2 to reduce the number of tests that need to be duplicated to support the react conversion and prioritize maintaining a full suite of tests that can support both backbone and react, while implementing best practices and reducing tech debt.
> 
> Option 1b chosen over 1a because it allows for prioritizing fixing the failing tests for the react signin full prod rollout.
> 

Reviewer 2: In this section perhaps we can rephrase to link the chosen decision more directly with the above criteria. Something like:

```suggestion
## Decision Outcome

Chosen option: **Option 1b, Single set of POMs for React/Backbone, update existing tests with a test-first approach**

Overall, option 1b best fulfills our time & effort decision criteria, keeping the signin React rollout on track. While the action plan outlined in option 2 has served the team well thus far, separating the React and Backbone concerns, it will not scale in the signin use case, where the POM and test case duplication required would considerably extend the timeline for releasing the React signin feature set. 

All options will maintain the same amount of test coverage, however option 1 better assures that requirement changes are universally applied to test cases for React and Backbone.

Option 1a offers the most clarity and efficiency in ensuring best practices and minimizing maintenance costs while moving towards are ideal POM state, but it's lack of focus on the blockers for the React signin rollout risks causing delays due to scope creep.

```

Please feel free to change/expand/ignore

# Improving functional test architecture to ease support for React flows

- Status: Draft
- Deciders: Katrina Anderson, Lauren Zugai, Valerie Pomerleau

> - Date: 20024-04-22

Reviewer 3: ```suggestion
- Date: 2024-04-22
```

### Previous react rollouts

- ResetPassword: Fully rolled out in production, backbone routes no longer used. All backbone tests are skipped. Only the new `react-conversion` tests are running. These tests were relatively well isolated and the migration strategy worked well.
- Signup: Full prod rollout was previously attempted and rolled back, second full prod rollout is planned for Train 283 (stage on 2024/04/24, production on 2024/05/01). New react tests were created, and where the signup flow was used in non-signup tests a condition was added with test steps to go through either flow based on feature flag. In preparation for the initial full prod rollout, backbone signup tests were skipped when the feature flag was on. By consequence, with the feature flag currently on for all environments, all backbone tests are skipped even though 85% of users see the backbone version of the signup flow.

### Migration strategy and challenges

Up until now, the primary strategy for supporting react flows with functional tests has mainly been to:

- create new react-specific POMs (page object models)

> - duplicate tests and place them in a `react-conversion` directory (tests selected for duplication are based on the [Playright Migration Analysis][2])

Reviewer 3: ```suggestion
- duplicate tests and place them in a `react-conversion` directory (tests selected for duplication are based on the [Playwright Migration Analysis][2])
```

##### Pros

- Organizes the work on a per-page basis, providing clear boundaries for updates.
- By focusing on updating POMs for individual pages, we can prioritize and tackle updates in a more granular manner. This approach allows for targeted modifications, ensuring that each page's functionality is properly handled.
- Updating POMs page by page minimizes the risk of disrupting dependencies across different pages. This incremental approach reduces the likelihood of introducing regressions and simplifies troubleshooting if issues arise.
- With updates concentrated on specific pages, testing efforts can be more focused and efficient. We can test changes for each page individually, ensuring that the updates meet the required criteria before moving on to the next page.
- As the project progresses, the page-first approach can scale more effectively to accommodate new pages or updates. Each new page can be integrated into the unified POM framework systematically, maintaining consistency across the test suite.
- And of course, less duplication

##### Cons

- Bad, because [argument b]

#### Option 1b: Unify Backbone & React POMS - Test First Approach

> This approach to unifying the POMs involves organizing the work on a per-test basis, and working from groups of tests and updating the corresponding POMs a needed.

Reviewer 4: ~a~ as needed
